### PR TITLE
Add configurable resolutions

### PR DESCRIPTION
**Pull request by westfeld**
I need some help where to put the file `resolutions.txt` in the build tree to be included in the `/boot` directory

adresses Issue #90

### CODE REVIEW

**Comment by htruong** [comment]
Straightforward and very clean. LGTM. @dgsiegel ?

**Comment by westfeld** [comment]
Thanks for the comment, still I need help how to place the `resolutions.txt` into `/boot`.

File: package/piwebcam/resolutions.txt
Comment (mseri) [inline package/piwebcam/resolutions.txt:3]: Shouldn’t this be “comment”?

Comment (westfeld) [reply package/piwebcam/resolutions.txt]: Yes, you are right - commited a fix in 19dc199

**Comment by westfeld** [comment]
Fixes #90

**Comment by dgsiegel** [comment]
Looks very nice, Thomas! Thanks for your work. I'll add a few comments to the patch directly, but would like to merge this as soon as possible and create a new release.

File: board/post-image.sh
Comment (dgsiegel) [inline board/post-image.sh:94]: I don't think this should be an optional file, like `camera.txt` is. We could certainly move this file to `/etc`.

In that case I *think* you could move that line to `package/piwebcam/piwebcam.mk`?

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [inline package/piwebcam/multi-gadget.sh:87]: See above, this could be a global configuration setting

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [inline package/piwebcam/multi-gadget.sh:70]: I am pondering about the right format of that file. One might want to use `raw` instead of `mjpeg` or even `mp4` if uvc-gadget gets support for that. In that case it would be better to add the format to that file as well. Also, it's quite error prone to parse it like that.

How about:

```
mjpeg 640 360
mjpeg 640 480
...
```
That way you could parse it with awk and drop that line if it does not contain out of 3 words. (Bonus point if you check for valid formats and if the latter two arguments are integers)

This would then for example allow to easily use `raw` etc. As for the configfs name you'd then just use the first letter of the format.

Further opinions welcome!

File: board/post-image.sh
Comment (westfeld) [reply board/post-image.sh:94]: I could do that, however the aim of this PR was to make the output resolutions configurable without having to build the whole image.

One could put the default file into `/etc` and make it possible to overwrite it with another file in `/boot` however I see no advantage in there.

File: package/piwebcam/multi-gadget.sh
Comment (westfeld) [reply package/piwebcam/multi-gadget.sh:70]: I will have a look at this. I would then rename the input file from `resolutions.txt` to `output_formats.conf`.

File: board/post-image.sh
Comment (dgsiegel) [reply board/post-image.sh:94]: Well, we get a new build with each PR/commit so this shouldn't be an issue :)

The advantage would be that configuration settings shouldn't belong inside shell files. Furthermore /boot is writeable, whereas /etc is not.

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [reply package/piwebcam/multi-gadget.sh:70]: Let's keep the "txt" extensions as all the other files do use that too. Looking forward to your ideas.

File: package/piwebcam/video_formats.txt
Comment (dgsiegel) [inline package/piwebcam/video_formats.txt:3]: mp4 is not yet supported unfortunately, see https://github.com/peterbay/uvc-gadget/issues/8

Currently we only have `mjpeg` and `uncompressed`, see: https://github.com/raspberrypi/linux/blob/rpi-5.4.y/drivers/usb/gadget/function/uvc_configfs.c#L767

File: board/post-image.sh
Comment (dgsiegel) [inline board/post-image.sh:94]: This should be part of `package/piwebcam/piwebcam.mk`. Any reason why you chose `video_formats` instead of `resolutions`. Happy with both, it's just a longer name.

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [inline package/piwebcam/multi-gadget.sh:69]: Looks a lot better. Might want to add `^` to start at the beginning of a line and/or exclude lines starting with `#`. awk might be perfect for this but happy to keep it like this as well.

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [inline package/piwebcam/multi-gadget.sh:74]: "m" should be the first character of the format instead of being hardcoded

File: package/piwebcam/multi-gadget.sh
Comment (westfeld) [reply package/piwebcam/multi-gadget.sh]: resolved, see fb0418a

File: package/piwebcam/video_formats.txt
Comment (westfeld) [reply package/piwebcam/video_formats.txt]: I corrected that in fb0418a

File: board/post-image.sh
Comment (westfeld) [reply board/post-image.sh:94]: My idea of the fix was to be able to configure the piwebcam to ignore resolutions after the image has been burned to an SD card.

So if you see that the resolution is too low, you can put the SD card into your PC and change the `video_formats.txt` file on the `/boot` partition, reinsert it back into your raspberrypi and it will work.

The reason why I changed the name was that now not only the resolution gets configurable, but also the streaming format. So I thought naming the file just `resolutions.txt` is misleading.

File: package/piwebcam/multi-gadget.sh
Comment (westfeld) [reply package/piwebcam/multi-gadget.sh:74]: To be honest I do not know what the "m" parameter in the config file function call does. It creates some file for USB configuration. So does it have to be changeable? How should the config file look like?

    m mjpeg 1920 1080

or

    mmjpeg 1920 1080

In the second case I would just take the very first character. What about the syntax checking? Should it be only "m" at the moment?

File: package/piwebcam/multi-gadget.sh
Comment (westfeld) [reply package/piwebcam/multi-gadget.sh]: To be honest, I have not dived into awk, yet. If it is really beneficial I would go that route, though.

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [reply package/piwebcam/multi-gadget.sh:74]: That single character is the specific streaming header descriptor, "m" for mjpeg and "u" for uncompressed. I *think* you could use any letter, hence I've said to just use the first letter of the format.

File: board/post-image.sh
Comment (dgsiegel) [reply board/post-image.sh:94]: Right, this approach is sensible. What I meant however was to use the Makefile instead of the shell script to add that file.

Comment (westfeld) [reply board/post-image.sh:94]: I have looked into this a bit more. But I did not find out how to place files in the `/boot` partition out of a buildroot package. I understood how to place files into the staging area to be included in the image, but I am not sure if this counts for the `/boot` partition, as well...

Comment (dgsiegel) [reply board/post-image.sh:94]: Oh you don't have to, /etc would be the right place for this. A user should then be able to overwrite that file from /boot. Just add this line to `package/piwebcam/piwebcam.mk`

```
$(INSTALL) -D -m 644 $(PIWEBCAM_PKGDIR)/video_formats.txt $(TARGET_DIR)/etc/video_formats
```

Nevertheless I think we should come up with a better name for this file?

Comment (westfeld) [reply board/post-image.sh]: I have implemented the installation in 618e516 and I am open to suggestions for another file name.

**Review by dgsiegel — DISMISSED** [review]

**Comment by dgsiegel** [comment]
@htruong any idea for a better file name? Other than that I think we can merge this now.

**Comment by westfeld** [comment]
What do you think of ’vformats.txt’

**Comment by htruong** [comment]
I think `video_formats.txt` is good (`vformats.txt`, on the other hand is short but is a bit obscure). I will give it a spin before merging :)

**Comment by westfeld** [comment]
I thought adding some documentation would be nice. Is it OK to put it there or would it be better in the wiki?

**Comment by dgsiegel** [comment]
> I thought adding some documentation would be nice. Is it OK to put it there or would it be better in the wiki?

This is great, thanks! I'd only add that client apps will chose resolutions automatically, hence it's only needed to reduce available resolutions if you want to enforce a specific setting.

**Comment by westfeld** [comment]
@dgsiegel Thanks for the kind introduction and the comments. It helped me as a first-contributor to this project. I am using the showmewebcam everyday and it is great (Pi0 + HQ).

**Comment by dgsiegel** [comment]
> @dgsiegel Thanks for the kind introduction and the comments. It helped me as a first-contributor to this project. I am using the showmewebcam everyday and it is great (Pi0 + HQ).

Thanks :) Glad to have you on board. Let's wait for @htruong to give it a spin and then merge it.

File: README.md
Comment (dgsiegel) [inline README.md:164]: Just a minor nitpick: Many haven't had this issue with Teams on WIndows, some did however. I'd guess this paragraph would be better suited for the FAQ (https://github.com/showmewebcam/showmewebcam/wiki/FAQ) than the Readme.

Comment (westfeld) [reply README.md]: I have moved the use case from the README to the wiki.

File: package/piwebcam/multi-gadget.sh
Comment (wherget) [inline package/piwebcam/multi-gadget.sh:88]: AFAICS, you'd need to link `functions/uvc.usb0/streaming/uncompressed/u` as well for uncompressed formats to work. Did you try it with `uncompressed`?

File: package/piwebcam/multi-gadget.sh
Comment (wherget) [inline package/piwebcam/multi-gadget.sh:72]: This sounds the wrong way round. If user file does NOT exist, use user file as formats file? Did you check this with a custom file that does specify some other formats?

Also, how do you feel about a stronger check, i.e. "file is readable" `-r` or even "not empty" `-s`

File: package/piwebcam/piwebcam.mk
Comment (wherget) [inline package/piwebcam/piwebcam.mk:24]: If only for symmetry reasons, I'd prefer the full set of mode bits, i.e. `0644`.

**Review by wherget — COMMENTED** [review]
I really like this (tbh I implemented almost the same, but yours has a few more niceties), but I think there's still a few kinks in the code, which I've taken the liberty to annotate.

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [reply package/piwebcam/multi-gadget.sh:88]: Oh true, I've missed that. Thanks!

@westfeld, basically for every given format you need to link the headers

For mjpeg:

`   ln -s functions/uvc.usb0/streaming/mjpeg/m  functions/uvc.usb0/streaming/header/h `

For uncompressed:

`   ln -s functions/uvc.usb0/streaming/uncompressed/u  functions/uvc.usb0/streaming/header/h `

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [reply package/piwebcam/multi-gadget.sh:72]: Right, you could set ` VIDEO_FORMATS_FILE=/etc/video_formats.txt` as default, and overwrite it if a user file exists. It should always be readable, as /boot is FAT. But you're right, better safe than sorry :)

**Review by htruong — COMMENTED** [review]
Also, can you kindly squash those commits together and then force push to the PR :)? That will make it a little bit cleaner.

**Comment by westfeld** [comment]
I have implemented the changes locally, but I have to test the resolutions and the handling of the uncompressed video before pushing, first.

**Comment by HunterAP23** [comment]
Saw that the issue I created was tagged in this, thanks for working so hard on this!

I wanted to ask if higher than 1080p resolutions were tested, as well as >30fps was also tested (or if they need to be tested). I know that #72 is a not-yet merged request or adding 60fps support, but I was curious to see if it was at least technically feasible. 

The Pi Zero / Zero W could theoretically support 1080p60 over USB 2 (30fps is 1.87 Gbps and 60fps would be 3.73 Gbps, which is below the 5Gbps spec of USB 2).

The Pi 4 has the advantage of having 10Gbps bandwidth, so it could achieve 3840x2160 at 30fps using 7.46 Gbps, or an odd resolution like 1440p at 60fps using 6.64 Gbps. 2160p would require the HQ camera module, whereas 1440p would requirethe Camera V2 module.

I can try and test these settings myself and reporting back if need be, I was just curious if there are plans to add those options or if there are any known issues for them

**Comment by westfeld** [comment]
Thanks @HunterAP23 for commenting. I will get back to you once I get this PR up and running.

Right now I have a HQ camera and a Pi zero and a Pi zero w for testing.

I will test several resolutions and streaming formats (mjpeg vs. uncompressed).

Changing the frame rate is at the moment not yet implemented, but it is certainly possible. I think it is important to only allow settings which are compatible to the used hardware.

**Review by htruong — APPROVED** [review]

File: package/piwebcam/multi-gadget.sh
Comment (wherget) [inline package/piwebcam/multi-gadget.sh:92]: Shellcheck reasonably complained about quoting here
```suggestion
    ln -sf "functions/uvc.usb0/streaming/$VIDEO_FORMAT/$HDR_DESC"  functions/uvc.usb0/streaming/header/h
```

**Review by htruong — COMMENTED** [review]
Squash with previous commit?

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [inline package/piwebcam/multi-gadget.sh:75]: Small nitpick: the file uses 2 spaces indentation, here's 4.

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [inline package/piwebcam/multi-gadget.sh:85]: Small nitpick: single space before `grep`

File: package/piwebcam/multi-gadget.sh
Comment (dgsiegel) [inline package/piwebcam/multi-gadget.sh:92]: Small nitpick: double space between the two arguments

File: package/piwebcam/video_formats.txt
Comment (dgsiegel) [inline package/piwebcam/video_formats.txt:16]: Small nitpick: Empty line

**Review by dgsiegel — CHANGES_REQUESTED** [review]
Just a few small nitpicks here and there. Other than that I am looking forward to our next release :)

**Comment by whi-tw** [comment]
Is there any way to speed this along if it appears to just be blocked by style changes (ie. fix the formatting during the merge process?)

**Comment by HunterAP23** [comment]
I wasn't sure how to go about making the changes requested, so I opted to make a fork of [westfeld:select_resolutions](https://github.com/westfeld/showmewebcam/tree/select_resolutions) with those changes made and created a [PR ](https://github.com/westfeld/showmewebcam/pull/1) to it. Once that PR is merged, those changes should appear on this PR.

**Comment by dgsiegel** [comment]
I've updated and rebased this PR in #154

**Comment by dgsiegel** [comment]
Merged via #154. Thanks @westfeld, @HunterAP23 & @wherget for your help!

**Pull request closed without merging**
